### ISSUE TITLE
autograd

### ISSUE THREAD

**Issue by viridyu** [DLHacks/pix2pix_PAN#1]
HI,

I have a question about the intermediate_output in PAN. Why you make the requires_grad=false in self.intermediate_outputs.append(Variable(output.data, requires_grad=False)) ? I think it might need the backward propagation of pan loss in PAN. Thank you.

**Comment by OtsuboAtsushi** [comment]
The intermediate_output is an output of each Conv2d (intermediate_conv). So it is not a parameter and we don't need to update it. The post below may be the same situation. https://discuss.pytorch.org/t/whether-variable-requires-grad/900

**Comment by viridyu** [comment]
@OtsuboAtsushi , Thank you for your answer. I think our situation may be different from the example you give. The features we extract are used to calculate in the pan loss. I mean we need the intermediate features to help back-propagate the grad to the weights of the net. If we set the it as not autograd, can the pan loss still influence the discriminator and generator?

**Comment by viridyu** [comment]
@OtsuboAtsushi , sorry, I may misunderstand how to use requires_grad. I also have a question about the def backward_D() and def backward_G(). If you use fake_AB.detach() in backward_D(), should it renew self.loss_PAN with the fake_AB in backward_G() ?

**Comment by OtsuboAtsushi** [comment]
@viridyu  Without .detach() in backward_D(), the G's parameters are updated in backward_D() because the fake_AB are connected (attached) to the G's computational graph. Please see https://discuss.pytorch.org/t/whats-the-difference-between-variable-detach-and-variable-clone/10758 .

> should it renew self.loss_PAN with the fake_AB in backward_G() ?

The loss_PAN is in common with D and G. In other words, D and G share the loss_PAN in each iteration. Please check in `def optimize_parameters(self)`, backward_D() is called first (loss_PAN is updated here) and then backward_G() are called (G uses the same loss_PAN).

**Comment by OtsuboAtsushi** [comment]
The original pix2pix code which I use as a reference is here https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/pix2pix_model.py .

**Comment by viridyu** [comment]
@OtsuboAtsushi  The original pix2pix uses the fake_AB ( not detached) to forward-pass discriminator again for G_GAN_loss.  if you use fake_AB.detach() to calculate self.loss_PAN in backward_D(), the generator's graph would not be connected to self.loss_PAN. Then, in backward_G(), the same self.loss_PAN cannot update generator's weights. So, I suggest to code like this:

`

def backward_G(self):
        
        fake_AB = torch.cat((self.real_A, self.fake_B), 1)
        pred_fake = self.netD.forward(fake_AB)
        fake_inters = self.netD.get_intermediate_outputs()

        self.loss_G_GAN = self.criterionGAN(pred_fake, True)

        real_AB = torch.cat((self.real_A, self.real_B), 1)
        pred_real = self.netD.forward(real_AB)
        real_inters = self.netD.get_intermediate_outputs()

        self.loss_PAN = 0
        for (fake_i, real_i, lam) in zip(fake_inters, real_inters, self.pan_lambdas):
            self.loss_PAN += self.criterionPAN(fake_i, real_i.detach()) * lam

        self.loss_G = self.loss_G_GAN + self.loss_PAN
        self.loss_G.backward()`

**Comment by OtsuboAtsushi** [comment]
Thank you. That's may be correct. I used old code as a reference. I'll fix it.
